File: Bronze/2020/US Open/03_cowntact_tracing/CowntactTracing.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.StringTokenizer;

/**
 * @author Red
 * @version 1.0
 */
public class CowntactTracing {
    static boolean[] cowEndsInfected;
    static int n;
    static int maxT = 250;
    static int maxN = 100;
    static int[] cowX = new int[maxT + 1];
    static int[] cowY = new int[maxT + 1];

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("tracing.in"));
        PrintWriter pw = new PrintWriter("tracing.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        n = Integer.parseInt(st.nextToken());
        int t = Integer.parseInt(st.nextToken());
        // 记录最终感染结果
        String s = r.readLine();
        cowEndsInfected = new boolean[n + 1];
        for (int x = 1; x <= n; x++) {
            cowEndsInfected[x] = (s.charAt(x - 1) == '1');
        }
        // 记录接触信息
        for (int i = 0; i < t; i++) {
            st = new StringTokenizer(r.readLine());
            int time = Integer.parseInt(st.nextToken());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            cowX[time] = x;
            cowY[time] = y;
        }
        // 枚举所有牛作为 patient zero 以及握手次数 k
        int lowerK = maxT + 1;
        int upperK = 0;
        HashSet<Integer> candidatePatientZero = new HashSet<>();
        for (int i = 1; i <= n; i++) {
            for (int k = 0; k <= maxT + 1; k++) {    // k 可以取 maxT+1（251）时说明最终为 Infinity
                if (consistentWithData(i, k)) {
                    lowerK = Math.min(lowerK, k);
                    upperK = Math.max(upperK, k);
                    candidatePatientZero.add(i);
                }
            }
        }
        pw.print(candidatePatientZero.size() + " " + lowerK + " ");
        if (upperK == maxT + 1) {
            pw.println("Infinity");
        } else {
            pw.println(upperK);
        }
        pw.close();
    }

    /**
     * 判断 patientZero 为传播起点握手 k 次 是否能满足最终感染结果
     */
    static boolean consistentWithData(int patientZero, int k) {
        boolean[] infected = new boolean[maxN + 1];
        int[] numHandshakes = new int[maxN + 1];
        infected[patientZero] = true;
        for (int t = 0; t <= maxT; t++) {
            int x = cowX[t];
            int y = cowY[t];
            if (x > 0) {
                if (infected[x]) {
                    numHandshakes[x]++;
                }
                if (infected[y]) {
                    numHandshakes[y]++;
                }
                if (numHandshakes[x] <= k && infected[x]) {
                    infected[y] = true;
                }
                if (numHandshakes[y] <= k && infected[y]) {
                    infected[x] = true;
                }
            }
        }
        for (int i = 1; i <= n; i++) {
            if (infected[i] != cowEndsInfected[i]) {
                return false;
            }
        }
        return true;
    }
}

File: Bronze/2020/US Open/03_cowntact_tracing/cowntact_tracing.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_set>

using namespace std;

const int MAX_T = 250;
const int MAX_N = 100;

bool cow_ends_infected[MAX_N + 1];
int n, cow_x[MAX_T + 1], cow_y[MAX_T + 1];
// 判断 patient_zero 为传播起点握手 k 次 是否能满足最终感染结果
bool consistent_with_data(int patient_zero, int k) {
    bool infected[MAX_N + 1] = {false};
    int num_handshakes[MAX_N + 1] = {0};
    infected[patient_zero] = true;
    for (int t = 0; t <= MAX_T; t++) {
        int x = cow_x[t];
        int y = cow_y[t];
        if (x > 0) {
            if (infected[x]) {
                num_handshakes[x]++;
            }
            if (infected[y]) {
                num_handshakes[y]++;
            }
            if (num_handshakes[x] <= k && infected[x]) {
                infected[y] = true;
            }
            if (num_handshakes[y] <= k && infected[y]) {
                infected[x] = true;
            }
        }
    }
    for (int i = 1; i <= n; i++) {
        if (infected[i] != cow_ends_infected[i]) {
            return false;
        }
    }
    return true;
}

int main() {
    ifstream fin("tracing.in");
    ofstream fout("tracing.out");

    fin >> n;
    int t;
    fin >> t;
    // 记录最终感染结果
    string s;
    fin >> s;
    // 记录接触信息
    for (int x = 1; x <= n; x++) {
        cow_ends_infected[x] = (s[x - 1] == '1');
    }
    // 枚举所有牛作为 patient zero 以及握手次数 k
    unordered_set<int> candidate_patient_zero;
    int lower_k = MAX_T + 1;
    int upper_k = 0;

    for (int i = 0; i < t; i++) {
        int time, x, y;
        fin >> time >> x >> y;
        cow_x[time] = x;
        cow_y[time] = y;
    }

    for (int i = 1; i <= n; i++) {
        // k 可以取 MAX_T + 1（251）时说明最终为 Infinity
        for (int k = 0; k <= MAX_T + 1; k++) {
            if (consistent_with_data(i, k)) {
                lower_k = min(lower_k, k);
                upper_k = max(upper_k, k);
                candidate_patient_zero.insert(i);
            }
        }
    }

    fout << candidate_patient_zero.size() << " " << lower_k << " ";

    if (upper_k == MAX_T + 1) {
        fout << "Infinity" << endl;
    } else {
        fout << upper_k << endl;
    }

    fin.close();
    fout.close();

    return 0;
}
